docs/proposals/container-runtime-interface/capabilities.md

```diff
@@ -0,0 +1,137 @@
+# Linux Capabilities in Container Runtime Interface
+
+The following proposal summarizes the current state and problems for runtimes to implement Linux Capabilities.
+
+## Problems
+
+Today's [Linux Capabilities Interface](https://github.com/kubernetes/kubernetes/blob/v1.5.0-alpha.0/pkg/kubelet/api/v1alpha1/runtime/api.proto#L350-L355) in CRI is derived from the Docker runtime implementation, which also maps
+to the [k8s API](https://github.com/kubernetes/kubernetes/blob/v1.5.0-alpha.0/pkg/api/v1/types.go#L1164-L1170).
+
+If users want to launch a container with ad-hoc capabilities, they will go through the following steps:
+
+- Get a list of capabilities he would require
+- Look up Docker's docs about the [default capability list](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities).
+- Calculate the `Add` and `Drop` capability set based on his requirements and the default capabilities list.
+- Specify the `Add` and `Drop` set for the containers in a k8s pod spec.
+- Launch the pod/container.
+
+There are several issues here:
+
+- The default capability set heavily depends on Docker's implementation, Kubernetes has no control over that.
+When Docker changes the default capability set across versions, the effective set of default capabilities can change suddenly.
+- To implement the same semantics as today's Docker runtime, other runtimes (e.g. rkt) are required to depend on the implicit Docker default capability set as well. And if Docker changes the default capability set, then consistency is not guaranteed.
+- From a security point of view, an explicit capability set which does not rely on undefined defaults is preferable for auditing purposes.
+
+## Solution
+
+The proposed solution will:
+- Keep the k8s API as is today. Users will not need to change anything.
+- Define a default capability whitelist inside k8s, and document it.
+- Add a `DefaultCapabilities` field to the `PodSecurityPolicySpec` which overrides the default capability whitelist inside k8s.
+- Set the default capability whitelist in the `PodSpec`'s `SecurityContext`.
```

Comment:
Going forward, ideally the author would just provide the complete (resolved) whitelist, but we need to consider the following workflows for backwards compatibility:

1. Only Add/Drop:
  - [user] Create Pod specifying only Add/Drop capabilities
  - [system] Resolve whitelist: Default (from system / PSP) + Add - Drop; set the whitelist in the PodSpec (default)
2. Only set whitelist:
  - [user] Create Pod specifying only the whitelist
  - [system] No action necessary.
3. Both a whitelist & Add/Drop
  - [user] Create or edit a Pod, with both Add/Drop & whitelist specified
  - [system] Resolve whitelist: specified whitelist + Add - Drop; Update the PodSpec as needed

Does that clear things up? Overall, the idea is that the "Whitelist" is the "correct" way going forward, but we need to continue to "resolve" it for backwards compatibility.